[#603][8] Add result panel component

henrychen0220 opened the pull request:
This PR adds a new functionality to display workflow results after execution to the new GUI.

Whenever an execution finishes, `ResultPanelComponent` will fetch the execution result and display it onto the bottom area of the new GUI. If the execution result shows that the workflow graph executes correctly, a new data table showing the execution result in excel table format will be displayed. Otherwise, when a error message is returned instead, the result panel will show the corresponding error message on the result panel.

Clicking on each row of the result table will create a pop-up window and display the detail of that row in the pretty json format. The pop-up window is implemented using the `modal` component provided by the `ng-bootstrap` library. 

A `paginator` is also added to the new GUI to prevent displaying too many results all at once in the small area used by the result panel.

**New result panel data table:**

![capture](https://user-images.githubusercontent.com/19577058/41278320-7910f2d4-6dde-11e8-85dd-362763c6f32e.JPG)


**New result panel error message:**

![capture1](https://user-images.githubusercontent.com/19577058/41278333-7fbf0828-6dde-11e8-9a74-553527ac253a.JPG)


**New result panel row detail:**

![capture2](https://user-images.githubusercontent.com/19577058/41278341-81fd3e84-6dde-11e8-8fe0-7fe491f32cc0.JPG)



`ng-bootstrap`: https://github.com/ng-bootstrap/ng-bootstrap 

This PR adds 5 new test cases.

avinash0161 commented on core/new-gui/src/app/workspace/component/operator-panel/operator-panel.component.spec.ts:
```diff
@@ -12,11 +13,17 @@ import { StubOperatorMetadataService } from '../../service/operator-metadata/stu
 
 import {
-  getMockOperatorMetaData, getMockOperatorGroup, getMockOperatorSchemaList
+  mockOperatorMetaData, mockOperatorGroup, mockOperatorSchemaList
```
Haven't these changes already gone through? #619

avinash0161 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -1,15 +1,169 @@
+   * @param content
+   */
+  public open(content: any): void {
+    this.modalService.open(content).result.then((result) => {
```
How can this function accept an argument which is in constructor but wasn't assigned to any local data member?

henrychen0220 replied:
`content` is a `TemplateRef` defined by the line `<ng-template #content let-c="close" let-d="dismiss">`, as described in the ng-bootstrap library: https://ng-bootstrap.github.io/#/components/modal/api.

avinash0161 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -1,15 +1,169 @@
+   * @param row
+   * @param content
+   */
+  public getRowDetails(row: any, content: any): void {
```
Didn't we disallow the use of 'any'?

zuozhiw replied:
Same comment, we do not allow `any` unless there's absolute no workaround. This is a strict rule. Please change this function and the function `open` above with property type signature.

henrychen0220 replied:
I have edited the type for `content` to `TemplateRef<any>` as described in this article,  https://blog.angular-university.io/angular-ng-template-ng-container-ngtemplateoutlet/.

As for the `row`, it represents each of the result instance fetched from backend, which have no defined format. Therefore, I kept it as type `any`.

avinash0161 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -1,15 +1,169 @@
+
+    if (resultData !== undefined) {
+      // generate columnDef from first row, column definition is in order
+      this.currentDisplayColumns = Object.keys(resultData[0]).filter(x => x !== '_id');
```
Please provide a brief overview in comments about what is being done here. An example of what comes as response and the transformation done would be great.

henrychen0220 replied:
Some additional comments have been added in the new commits.

avinash0161 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -1,15 +1,169 @@
+/**
+ * Class for holding the properties for each column
+ */
+export class TableColumn {
```
This should go in the folder holding all the interfaces. I think it's types.

henrychen0220 replied:
New commit has implemented `TableColumn` as interface as suggested.

avinash0161 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.html:
```diff
@@ -1,5 +1,33 @@
+      </ng-container>
+      <mat-header-row *matHeaderRowDef="currentDisplayColumns"></mat-header-row>
+      <mat-row class="result-table-row" *matRowDef="let row; columns: currentDisplayColumns;"
+          (click)="getRowDetails(row, content)"></mat-row>
```
Where does this content in (row, content) come from? I don't see it being initialized anywhere above.

henrychen0220 replied:
`content` is defined at the line `<ng-template #content let-c="close" let-d="dismiss">`.
This part of the code is straightly provided by the ng-bootstrap library.

https://ng-bootstrap.github.io/#/components/modal/examples

henrychen0220 commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -1,15 +1,159 @@
+  public currentDataSource: MatTableDataSource<object> | undefined;
+  public currentDisplayRow: string = '';
+
+  @ViewChild(MatPaginator) paginator: MatPaginator | null = null;
```
For this, apparently, MatTableDataSource.paginator has been set as `MatPaginator | null` by the angular material.

![capture](https://user-images.githubusercontent.com/19577058/41893096-6a7f7ed8-78cf-11e8-8a79-c55505ab2fdd.PNG)

Changing the type to `MatPaginator | undefined` will cause an error to the code.

zuozhiw replied:
Then there's nothing we can do... Sadly someone in the Angular team who wrote this code doesn't agree with me on the `null` and `undefined` thing

henrychen0220 commented:
New changes have been pushed to this PR.

Changes such as:

1. Moved ng-bootstrap modal to separate html document as Zuozhi suggested. With this, we can prevent the use of TemplateRef type while making the code much easier to understand. 

2. Added new interface called `IndexableObject`. Since in typescript, type `object` is not indexable, which we couldn't do operation such as `row[col]` where row is type `object`, we need to make such object indexable. Therefore, I create a new interface that basically makes `object` indexable using the `index signature` defined in `https://www.typescriptlang.org/docs/handbook/interfaces.html`.

3. Used [hidden] attribute rather than *ngIf directive to hide the result table when there is no result. The reason behind this change is that, when *ngIf is false, the element inside the *ngIf clause does not exist. When a new successful result that has more than 10 results come in, it will require paginator. However, since paginator is fetched using ViewChild(), by this time, <mat-paginator> will ceased to exist since *ngIf still haven't update the GUI. Therefore, `this.paginator` will be `undefined` and the paginator for the data table will not work as expected. On the other hand, if we use `[hidden]` attribute, the element still exist in the GUI. This only hides the elements by changing their css display values. Thus, `this.paginator` can fetch the correct paginator from the HTML.

4. Fixed 1 minor bug when the execution result successes but have 0 result (0 in view result).

zuozhiw commented:
@henrychen0220 the new changes look great! Good job. I think using the component with the modal makes the code look less magic and more readable.

Here's my second round of review:
1. I made some new minor changes, mostly modifying some comments. I also did a little refactor, I created two new functions, `displayErrorMessage`, and `displayResultTable`, to replace the previous one function that handles both cases. You can check them in the code.

Previously, when the backend returns an error, only `showMessage` is set to true, but other variables are not set to `undefined`. The problem with this is that the data inside those variables cannot be garbage collected, they are still holding the old data until the next success execution.

2. Please add one more test case to test when the result size is 0.

3. Regarding the `hidden` directive instead of `ngIf`, it seems like `ngIf` would cause a problem. However, I still think `ngIf` is the correct way. The difference is that `ngIf` will actually destroy the DOM element when the condition is false, `hidden` will not destroy the DOM element, it will just hide it. IMO it's still better to let `ngIf` destroy the DOM element completely when we don't show the results.

Therefore it's still better to use `ngIf`. I'm testing the code on my own and I changed `[hidden]` to `ngIf` (and reverse the condition). I tested it multiple times and it works fine for me. How can I reproduce the problem if we use `ngIf` ?

zuozhiw commented:
@henrychen0220 I'm able to reproduce the problem. I think one potential way to deal with this is that we manually call Angular's change detection function after we initialize the table variables and display the table, and then the child component will initialize and then we can pass in the paginator. 

But I think that solution again adds unnecessarily complications to the code. So think twice I feel that `hidden` is a better solution for now. Therefore I'm ok with it. 

BTW can you please add another test case: when the component is first created, before receiving any data, it shouldn't have anything displayed in it: the `message` div should not be there, and the `table` div should be hidden, maybe you can add this test in the `it should create` test case. Thanks.

zuozhiw commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.html:
```diff
@@ -1,5 +1,5 @@
 <div class="texera-workspace-result-panel-body texera-workflow-component-body">
-  <div *ngIf="showMessage">
+  <div class="error-message" *ngIf="showMessage">
```
change the class name to be a more specific name, prefix it with `texera-`, such as `texera-result-panel-message`
This is to minimize the conflict of name of our component with other library's name. Another library could also have a class name called "error-message".

zuozhiw commented:
@henrychen0220 one more minor comment about HTML class name, after that I approve this PR
